Evolution/0016-mutli-producer-single-consumer-channel.md

```diff
@@ -0,0 +1,847 @@
+            public static func watermark(low: Int, high: Int) -> BackpressureStrategy
+
+            /// A backpressure strategy using a high and low watermark to suspend and resume production respectively.
+            ///
+            /// - Parameters:
+            ///   - low: When the number of buffered elements drops below the low watermark, producers will be resumed.
+            ///   - high: When the number of buffered elements rises above the high watermark, producers will be suspended.
+            ///   - waterLevelForElement: A closure used to compute the contribution of each buffered element to the current water level.
+            ///
+            /// - Note, `waterLevelForElement` will be called on each element when it is written into the source and when
+            /// it is consumed from the channel, so it is recommended to provide a function that runs in constant time.
+            public static func watermark(low: Int, high: Int, waterLevelForElement: @escaping @Sendable (borrowing Element) -> Int) -> BackpressureStrategy
+
+            /// An unbounded backpressure strategy.
+            ///
+            /// - Important: Only use this strategy if the production of elements is limited through some other mean. Otherwise
+            /// an unbounded backpressure strategy can result in infinite memory usage and cause
+            /// your process to run out of memory.
+            public static func unbounded() -> BackpressureStrategy
+        }
+
+        /// A type that indicates the result of sending elements to the source.
+        public enum SendResult: ~Copyable, Sendable {
+            /// An opaque token that is returned when the channel's backpressure strategy indicated that production should
+            /// be suspended. Use this token to enqueue a callback by  calling the ``MultiProducerSingleConsumerAsyncChannel/Source/enqueueCallback(callbackToken:onProduceMore:)`` method.
+            ///
+            /// - Important: This token must only be passed once to ``MultiProducerSingleConsumerAsyncChannel/Source/enqueueCallback(callbackToken:onProduceMore:)``
+            ///  and ``MultiProducerSingleConsumerAsyncChannel/Source/cancelCallback(callbackToken:)``.
+            public struct CallbackToken: Sendable, Hashable { }
+
+            /// Indicates that more elements should be produced and send to the source.
+            case produceMore
+
+            /// Indicates that a callback should be enqueued.
+            ///
+            /// The associated token should be passed to the ````MultiProducerSingleConsumerAsyncChannel/Source/enqueueCallback(callbackToken:onProduceMore:)```` method.
+            case enqueueCallback(CallbackToken)
+        }
+
+        /// A callback to invoke when the channel finished.
+        ///
+        /// This is called after the last element has been consumed by the channel.
+        public func setOnTerminationCallback(_ callback: @escaping @Sendable () -> Void)
```

Comment:
it might be useful to have a parameter in the closure to tell why the termination happened, either finished, failed, or cancelled.